File: NOVEMBER 24/13 November/2563. Count the Number of Fair Pairs LC.cpp
class Solution {
public:
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        sort(nums.begin(), nums.end());
        size_t len = nums.size();
        long long pairCount = 0;
        for(int i=0; i<len ; i++) 
        {
            int lb = lower_bound(nums.begin()+i+1, nums.end(), lower-nums[i]) - nums.begin();
            int ub = lower_bound(nums.begin()+i+1, nums.end(), upper-nums[i]+1) - nums.begin();

            pairCount += (ub-lb);
        }
        // pairCount/=2;
        return pairCount;
    }
};

File: APRIL 2025/19 April/2563. Count the Number of Fair Pairs LC.java
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        Arrays.sort(nums);
        int N = nums.length;
        long countOfFairPairs = 0, lb, ub;
        for(int i=0; i<N; i++) {
            lb = (lower-nums[i]);
            ub = (upper-nums[i]);
            countOfFairPairs += (bound(nums, ub, i+1, N-1, true) - bound(nums, lb, i+1, N-1, false));
        }
        return countOfFairPairs;
    }

    private int bound(int[] nums, long target, int left, int right, boolean isUpperBound)
    {
        while(left<=right)
        {
            int mid = left+(right-left)/2;
            if(!isUpperBound) // when finding lower bound of target in nums[]
            {
                if(nums[mid]<target)
                    left = mid+1;
                else
                    right = mid-1;
            }
            else
            {
                if(nums[mid]<=target)
                    left = mid+1;
                else
                    right = mid-1;
            }
        }
        if(isUpperBound)
            return right;
        return left-1;
    }
}
